prow/spyglass/spyglass.go

```diff
@@ -144,101 +135,106 @@ func (s *Spyglass) Refresh(src string, podName string, sizeLimit int64, viewReq
+		return artifactNames, nil
+	default:
+		return nil, fmt.Errorf("Unrecognized key type for src: %v", src)
+	}
+}
+
+// prowToGCS returns the GCS key corresponding to the given prow key
+func (s *Spyglass) prowToGCS(prowKey string) (string, error) {
+	parsed := strings.Split(prowKey, "/")
+	if len(parsed) != 2 {
+		return "", fmt.Errorf("Could not get GCS src: prow src %q incorrectly formatted", prowKey)
+	}
+	jobName := parsed[0]
+	buildID := parsed[1]
+
+	job, err := s.jobAgent.GetProwJob(jobName, buildID)
+	if err != nil {
+		return "", fmt.Errorf("Failed to get prow job from src %q: %v", prowKey, err)
+	}
+
+	url := job.Status.URL
+	buildIndex := strings.Index(url, "/build/")
+	if buildIndex == -1 {
+		return "", fmt.Errorf("Unrecognized GCS url: %q", url)
 	}
-	return res
+	return url[buildIndex+len("/build/"):], nil
 }
 
 // FetchArtifacts constructs and returns Artifact objects for each artifact name in the list.
 // This includes getting any handles needed for read write operations, direct artifact links, etc.
 func (s *Spyglass) FetchArtifacts(src string, podName string, sizeLimit int64, artifactNames []string) ([]viewers.Artifact, error) {
-	foundArtifacts := []viewers.Artifact{}
-	foundArtifactNames := []string{}
-	for _, ep := range s.Eyepieces {
-		jobSource, err := ep.createJobSource(src)
-		if err == ErrCannotParseSource {
-			continue
-		} else if err != nil {
-			logrus.WithField("src", src).WithError(err).Error("Error creating job source.")
-			continue
-		}
-		artStart := time.Now()
+	artStart := time.Now()
+	arts := []viewers.Artifact{}
+	split := strings.SplitN(src, "/", 2)
+	keyType := split[0]
+	key := split[1]
+	switch keyType {
+	case "gcs":
 		for _, name := range artifactNames {
-			artifact := ep.artifact(jobSource, name, sizeLimit)
-			foundArtifacts = append(foundArtifacts, artifact)
-			foundArtifactNames = append(foundArtifactNames, artifact.JobPath())
+			art, err := s.GCSArtifactFetcher.artifact(key, name, sizeLimit)
+			if err != nil {
+				logrus.Errorf("Failed to fetch artifact %s: %v", name, err)
+				continue
+			}
+			arts = append(arts, art)
 		}
-		artElapsed := time.Since(artStart)
-		logrus.WithField("duration", artElapsed).Infof("Retrieved artifacts for %s", src)
-
-	}
-
-	// Special-casing the fetching of in-cluster artifacts that havent been found yet
-	neededArtifactNames := missingArtifacts(foundArtifactNames)
-	for _, artifactName := range neededArtifactNames {
-		switch artifactName {
-		case "build-log.txt":
-			if isProwJobSource(src) {
-				parsed := strings.Split(strings.TrimPrefix(src, "prowjob/"), "/")
-				var jobName string
-				var buildID string
-				if len(parsed) == 2 {
-					jobName = parsed[0]
-					buildID = parsed[1]
-				} else if len(parsed) == 1 {
-					podName = parsed[0]
-				} else {
-					return foundArtifacts, errors.New("invalid Prowjob source provided")
-				}
-				_, err := s.JobAgent.GetProwJob(jobName, buildID)
-				if err != nil {
-					continue
+	case "prowjob":
```

Comment:
Why would anyone ever use the gcs endpoint if this is a superset?

Reply:
GCS is stable, prowjob CRDs get deleted